File: platforms/dotnet/src/MLKemNative/MLKemNative768.cs
using System.Security.Cryptography;

namespace MLKemNative;

/// <summary>Base class for ML-KEM-768 API errors.</summary>
public abstract class MLKemException : CryptographicException
{
    private protected MLKemException(string message)
        : base(message)
    {
    }

    private protected MLKemException(string message, Exception innerException)
        : base(message, innerException)
    {
    }

    public sealed class InvalidPublicKey : MLKemException
    {
        public InvalidPublicKey(string message = "Invalid ML-KEM-768 public key")
            : base(message)
        {
        }
    }

    public sealed class InvalidPrivateKeyRepresentation : MLKemException
    {
        public InvalidPrivateKeyRepresentation(string message = "Invalid ML-KEM-768 private key representation")
            : base(message)
        {
        }
    }

    public sealed class InvalidCiphertext : MLKemException
    {
        public InvalidCiphertext(string message = "Invalid ML-KEM-768 ciphertext")
            : base(message)
        {
        }
    }

    public sealed class InvalidIncrementalHeader : MLKemException
    {
        public InvalidIncrementalHeader(string message = "Invalid ML-KEM-768 incremental header")
            : base(message)
        {
        }
    }

    public sealed class InvalidEncapsulationKeyVector : MLKemException
    {
        public InvalidEncapsulationKeyVector(string message = "Invalid ML-KEM-768 encapsulation key vector")
            : base(message)
        {
        }
    }

    public sealed class InvalidIncrementalEncapsulationSecret : MLKemException
    {
        public InvalidIncrementalEncapsulationSecret(
            string message = "Invalid ML-KEM-768 incremental encapsulation secret")
            : base(message)
        {
        }
    }

    public sealed class RandomGenerationFailed : MLKemException
    {
        public RandomGenerationFailed(Exception cause)
            : base("Secure random generation failed", cause)
        {
        }
    }

    public sealed class OperationFailed : MLKemException
    {
        public OperationFailed(string message = "ML-KEM-768 operation failed")
            : base(message)
        {
        }
    }
}

/// <summary>Pure managed ML-KEM-768 operations and byte-size constants.</summary>
public static class MLKemNative768
{
    public const int PublicKeyBytes = PureCSharpMLKEM768.PublicKeyBytes;
    public const int CiphertextBytes = PureCSharpMLKEM768.CiphertextBytes;
    public const int SharedSecretBytes = PureCSharpMLKEM768.SharedSecretBytes;
    public const int SecretKeyBytes = PureCSharpMLKEM768.SecretKeyBytes;
    public const int KeypairSeedBytes = PureCSharpMLKEM768.KeypairSeedBytes;
    public const int EncapsulationSeedBytes = PureCSharpMLKEM768.EncapsulationSeedBytes;
    public const int IncrementalHeaderBytes = PureCSharpMLKEM768.IncrementalHeaderBytes;
    public const int EncapsulationKeyVectorBytes = PureCSharpMLKEM768.EncapsulationKeyVectorBytes;
    public const int CiphertextPart1Bytes = PureCSharpMLKEM768.CiphertextPart1Bytes;
    public const int CiphertextPart2Bytes = PureCSharpMLKEM768.CiphertextPart2Bytes;
    public const int IncrementalEncapsulationSecretBytes =
        PureCSharpMLKEM768.IncrementalEncapsulationSecretBytes;
    public const int PrivateKeyRepresentationBytes = 5 + KeypairSeedBytes + PublicKeyBytes;

    private const int EkSeedOffset = EncapsulationKeyVectorBytes;
    private const int HeaderHashOffset = EncapsulationSeedBytes;
    private static readonly byte[] MagicBytes = { 0x4b, 0x4d, 0x4c, 0x4b, 0x31 };

    public sealed class PrivateKey
    {
        private readonly byte[] _secretKey;
        private readonly byte[] _representationBytes;

        private PrivateKey(byte[] secretKey, byte[] representationBytes, PublicKey publicKey)
        {
            _secretKey = secretKey;
            _representationBytes = representationBytes;
            PublicKey = publicKey;
        }

        public PublicKey PublicKey { get; }

        public byte[] Representation => Copy(_representationBytes);

        public static PrivateKey Generate()
        {
            byte[] seed = RandomBytes(KeypairSeedBytes);
            return FromSeed(seed);
        }

        public static PrivateKey FromRepresentation(byte[] representation)
        {
            ArgumentNullException.ThrowIfNull(representation);
            if (representation.Length != PrivateKeyRepresentationBytes || !StartsWithMagic(representation))
            {
                throw new MLKemException.InvalidPrivateKeyRepresentation();
            }

            int seedStart = MagicBytes.Length;
            int publicKeyStart = seedStart + KeypairSeedBytes;
            byte[] seed = Slice(representation, seedStart, publicKeyStart);
            byte[] expectedPublicKey = Slice(representation, publicKeyStart, PrivateKeyRepresentationBytes);
            PrivateKey generated = FromSeed(seed);

            if (!PureCSharpMLKEM768.ConstantTimeCompare(generated.PublicKey.RawRepresentation, expectedPublicKey))
            {
                throw new MLKemException.InvalidPrivateKeyRepresentation();
            }

            return generated;
        }

        internal static PrivateKey FromSeedForTesting(byte[] seed)
        {
            ArgumentNullException.ThrowIfNull(seed);
            if (seed.Length != KeypairSeedBytes)
            {
                throw new MLKemException.InvalidPrivateKeyRepresentation();
            }

            return FromSeed(Copy(seed));
        }

        public byte[] Decapsulate(byte[] ciphertext)
        {
            ArgumentNullException.ThrowIfNull(ciphertext);
            if (ciphertext.Length != CiphertextBytes)
            {
                throw new MLKemException.InvalidCiphertext();
            }

            return PureCSharpMLKEM768.Decapsulate(Copy(ciphertext), _secretKey);
        }

        public byte[] DecapsulateParts(byte[] ciphertextPart1, byte[] ciphertextPart2) =>
            MLKemNative768.DecapsulateParts(this, ciphertextPart1, ciphertextPart2);

        private static PrivateKey FromSeed(byte[] seed)
        {
            if (seed.Length != KeypairSeedBytes)
            {
                throw new MLKemException.OperationFailed("Invalid ML-KEM-768 keypair seed");
            }

            (byte[] publicKey, byte[] secretKey) = PureCSharpMLKEM768.KeypairDerand(seed);
            return BuildPrivateKey(seed, publicKey, secretKey);
        }

        private static PrivateKey BuildPrivateKey(byte[] seed, byte[] publicKey, byte[] secretKey)
        {
            if (!PureCSharpMLKEM768.CheckSecretKey(secretKey))
            {
                throw new MLKemException.OperationFailed("ML-KEM-768 secret key validation failed");
            }

            var representation = new byte[PrivateKeyRepresentationBytes];
            Array.Copy(MagicBytes, 0, representation, 0, MagicBytes.Length);
            Array.Copy(seed, 0, representation, MagicBytes.Length, KeypairSeedBytes);
            Array.Copy(publicKey, 0, representation, MagicBytes.Length + KeypairSeedBytes, PublicKeyBytes);

            return new PrivateKey(Copy(secretKey), representation, new PublicKey(publicKey));
        }
    }

    public sealed class PublicKey
    {
        private readonly byte[] _rawRepresentationBytes;

        public PublicKey(byte[] rawRepresentation)
        {
            ArgumentNullException.ThrowIfNull(rawRepresentation);
            _rawRepresentationBytes = Copy(rawRepresentation);
            if (_rawRepresentationBytes.Length != PublicKeyBytes ||
                !PureCSharpMLKEM768.CheckPublicKey(_rawRepresentationBytes))
            {
                throw new MLKemException.InvalidPublicKey();
            }
        }

        public byte[] RawRepresentation => Copy(_rawRepresentationBytes);

        public Encapsulation Encapsulate()
        {
            byte[] coins = RandomBytes(EncapsulationSeedBytes);
            return EncapsulateDerand(coins);
        }

        internal Encapsulation EncapsulateDerand(byte[] coins)
        {
            ArgumentNullException.ThrowIfNull(coins);
            if (coins.Length != EncapsulationSeedBytes)
            {
                throw new MLKemException.OperationFailed("Invalid ML-KEM-768 encapsulation seed");
            }

            (byte[] ciphertext, byte[] sharedSecret) =
                PureCSharpMLKEM768.EncapsulateDerand(_rawRepresentationBytes, Copy(coins));
            return new Encapsulation(ciphertext, sharedSecret);
        }
    }

    public sealed class Encapsulation : IEquatable<Encapsulation>
    {
        private readonly byte[] _ciphertextBytes;
        private readonly byte[] _sharedSecretBytes;

        internal Encapsulation(byte[] ciphertextBytes, byte[] sharedSecretBytes)
        {
            _ciphertextBytes = Copy(ciphertextBytes);
            _sharedSecretBytes = Copy(sharedSecretBytes);
        }

        public byte[] Ciphertext => Copy(_ciphertextBytes);

        public byte[] SharedSecret => Copy(_sharedSecretBytes);

        public bool Equals(Encapsulation? other) =>
            other is not null &&
            _ciphertextBytes.SequenceEqual(other._ciphertextBytes) &&
            _sharedSecretBytes.SequenceEqual(other._sharedSecretBytes);

        public override bool Equals(object? obj) => Equals(obj as Encapsulation);

        public override int GetHashCode() =>
            HashCode.Combine(ContentHashCode(_ciphertextBytes), ContentHashCode(_sharedSecretBytes));
    }

    public sealed class IncrementalPublicKey : IEquatable<IncrementalPublicKey>
    {
        private readonly byte[] _headerBytes;
        private readonly byte[] _encapsulationKeyVectorBytes;

        public IncrementalPublicKey(byte[] header, byte[] encapsulationKeyVector)
        {
            byte[] publicKey = PublicKeyFromIncremental(header, encapsulationKeyVector);
            _headerBytes = Copy(header);
            _encapsulationKeyVectorBytes = Copy(encapsulationKeyVector);
            PublicKey = new PublicKey(publicKey);
        }

        internal IncrementalPublicKey(byte[] header, byte[] encapsulationKeyVector, PublicKey publicKey)
        {
            _headerBytes = Copy(header);
            _encapsulationKeyVectorBytes = Copy(encapsulationKeyVector);
            PublicKey = publicKey;
        }

        public byte[] Header => Copy(_headerBytes);

        public byte[] EncapsulationKeyVector => Copy(_encapsulationKeyVectorBytes);

        public PublicKey PublicKey { get; }

        public bool Equals(IncrementalPublicKey? other) =>
            other is not null &&
            _headerBytes.SequenceEqual(other._headerBytes) &&
            _encapsulationKeyVectorBytes.SequenceEqual(other._encapsulationKeyVectorBytes);

        public override bool Equals(object? obj) => Equals(obj as IncrementalPublicKey);

        public override int GetHashCode() =>
            HashCode.Combine(ContentHashCode(_headerBytes), ContentHashCode(_encapsulationKeyVectorBytes));
    }

    public sealed class IncrementalEncapsulationPart1 : IEquatable<IncrementalEncapsulationPart1>
    {
        private readonly byte[] _encapsulationSecretBytes;
        private readonly byte[] _ciphertextPart1Bytes;
        private readonly byte[] _sharedSecretBytes;

        internal IncrementalEncapsulationPart1(byte[] encapsulationSecret, byte[] ciphertextPart1, byte[] sharedSecret)
        {
            _encapsulationSecretBytes = Copy(encapsulationSecret);
            _ciphertextPart1Bytes = Copy(ciphertextPart1);
            _sharedSecretBytes = Copy(sharedSecret);
        }

        public byte[] EncapsulationSecret => Copy(_encapsulationSecretBytes);

        public byte[] EncapsSecret => EncapsulationSecret;

        public byte[] CiphertextPart1 => Copy(_ciphertextPart1Bytes);

        public byte[] SharedSecret => Copy(_sharedSecretBytes);

        public bool Equals(IncrementalEncapsulationPart1? other) =>
            other is not null &&
            _encapsulationSecretBytes.SequenceEqual(other._encapsulationSecretBytes) &&
            _ciphertextPart1Bytes.SequenceEqual(other._ciphertextPart1Bytes) &&
            _sharedSecretBytes.SequenceEqual(other._sharedSecretBytes);

        public override bool Equals(object? obj) => Equals(obj as IncrementalEncapsulationPart1);

        public override int GetHashCode() => HashCode.Combine(
            ContentHashCode(_encapsulationSecretBytes),
            ContentHashCode(_ciphertextPart1Bytes),
            ContentHashCode(_sharedSecretBytes));
    }

    public static IncrementalPublicKey PublicKeyToIncremental(PublicKey publicKey)
    {
        ArgumentNullException.ThrowIfNull(publicKey);
        return PublicKeyToIncremental(publicKey.RawRepresentation);
    }

    public static IncrementalPublicKey PublicKeyToIncremental(byte[] publicKey)
    {
        PublicKey validatedPublicKey = new(publicKey);
        byte[] rawPublicKey = validatedPublicKey.RawRepresentation;
        byte[] header = new byte[IncrementalHeaderBytes];
        Array.Copy(rawPublicKey, EkSeedOffset, header, 0, EncapsulationSeedBytes);
        Array.Copy(PureCSharpMLKEM768.Sha3256Public(rawPublicKey), 0, header, HeaderHashOffset, SharedSecretBytes);

        return new IncrementalPublicKey(
            header,
            Slice(rawPublicKey, 0, EkSeedOffset),
            validatedPublicKey);
    }

    public static byte[] PublicKeyFromIncremental(byte[] header, byte[] encapsulationKeyVector)
    {
        ArgumentNullException.ThrowIfNull(header);
        ArgumentNullException.ThrowIfNull(encapsulationKeyVector);
        if (header.Length != IncrementalHeaderBytes)
        {
            throw new MLKemException.InvalidIncrementalHeader();
        }
        if (encapsulationKeyVector.Length != EncapsulationKeyVectorBytes)
        {
            throw new MLKemException.InvalidEncapsulationKeyVector();
        }

        var publicKey = new byte[PublicKeyBytes];
        Array.Copy(encapsulationKeyVector, 0, publicKey, 0, EncapsulationKeyVectorBytes);
        Array.Copy(header, 0, publicKey, EkSeedOffset, EncapsulationSeedBytes);

        byte[] expectedHash = PureCSharpMLKEM768.Sha3256Public(publicKey);
        byte[] headerHash = Slice(header, HeaderHashOffset, IncrementalHeaderBytes);
        if (!PureCSharpMLKEM768.ConstantTimeCompare(expectedHash, headerHash))
        {
            throw new MLKemException.InvalidIncrementalHeader("Invalid ML-KEM-768 incremental header hash");
        }

        if (!PureCSharpMLKEM768.CheckPublicKey(publicKey))
        {
            throw new MLKemException.InvalidPublicKey();
        }

        return publicKey;
    }

    public static IncrementalEncapsulationPart1 EncapsulatePart1(byte[] header, byte[]? randomness = null)
    {
        ArgumentNullException.ThrowIfNull(header);
        if (header.Length != IncrementalHeaderBytes)
        {
            throw new MLKemException.InvalidIncrementalHeader();
        }

        byte[] coins = randomness is null ? RandomBytes(EncapsulationSeedBytes) : Copy(randomness);
        if (coins.Length != EncapsulationSeedBytes)
        {
            throw new MLKemException.OperationFailed("Invalid ML-KEM-768 encapsulation randomness");
        }

        PureCSharpMLKEM768.IncrementalEncapsulationResult output =
            PureCSharpMLKEM768.EncapsulatePart1Derand(Copy(header), coins);

        return new IncrementalEncapsulationPart1(
            output.EncapsulationSecret,
            output.CiphertextPart1,
            output.SharedSecret);
    }

    public static byte[] EncapsulatePart2(
        byte[] encapsSecret,
        byte[] header,
        byte[] encapsulationKeyVector)
    {
        ArgumentNullException.ThrowIfNull(encapsSecret);
        if (encapsSecret.Length != IncrementalEncapsulationSecretBytes)
        {
            throw new MLKemException.InvalidIncrementalEncapsulationSecret();
        }

        byte[] publicKey = PublicKeyFromIncremental(header, encapsulationKeyVector);
        return PureCSharpMLKEM768.EncapsulatePart2(Copy(encapsSecret), publicKey);
    }

    public static byte[] DecapsulateParts(
        PrivateKey privateKey,
        byte[] ciphertextPart1,
        byte[] ciphertextPart2)
    {
        ArgumentNullException.ThrowIfNull(privateKey);
        ArgumentNullException.ThrowIfNull(ciphertextPart1);
        ArgumentNullException.ThrowIfNull(ciphertextPart2);
        if (ciphertextPart1.Length != CiphertextPart1Bytes)
        {
            throw new MLKemException.InvalidCiphertext("Invalid ML-KEM-768 ciphertext part1");
        }
        if (ciphertextPart2.Length != CiphertextPart2Bytes)
        {
            throw new MLKemException.InvalidCiphertext("Invalid ML-KEM-768 ciphertext part2");
        }

        var ciphertext = new byte[CiphertextBytes];
        Array.Copy(ciphertextPart1, 0, ciphertext, 0, CiphertextPart1Bytes);
        Array.Copy(ciphertextPart2, 0, ciphertext, CiphertextPart1Bytes, CiphertextPart2Bytes);
        return privateKey.Decapsulate(ciphertext);
    }

    internal static (byte[] PublicKey, byte[] SecretKey) KeypairDerandForTesting(byte[] seed)
    {
        ArgumentNullException.ThrowIfNull(seed);
        (byte[] publicKey, byte[] secretKey) = PureCSharpMLKEM768.KeypairDerand(Copy(seed));
        return (publicKey, secretKey);
    }

    internal static Encapsulation EncapsulateDerandForTesting(byte[] publicKey, byte[] coins) =>
        new PublicKey(publicKey).EncapsulateDerand(coins);

    private static byte[] RandomBytes(int count)
    {
        var output = new byte[count];
        try
        {
            RandomNumberGenerator.Fill(output);
        }
        catch (Exception exception)
        {
            throw new MLKemException.RandomGenerationFailed(exception);
        }

        return output;
    }

    private static bool StartsWithMagic(byte[] bytes)
    {
        if (bytes.Length < MagicBytes.Length)
        {
            return false;
        }

        for (int i = 0; i < MagicBytes.Length; i++)
        {
            if (bytes[i] != MagicBytes[i])
            {
                return false;
            }
        }

        return true;
    }

    private static byte[] Copy(byte[] source)
    {
        var result = new byte[source.Length];
        Array.Copy(source, result, source.Length);
        return result;
    }

    private static byte[] Slice(byte[] source, int start, int end)
    {
        var result = new byte[end - start];
        Array.Copy(source, start, result, 0, end - start);
        return result;
    }

    private static int ContentHashCode(byte[] bytes)
    {
        var hash = new HashCode();
        foreach (byte value in bytes)
        {
            hash.Add(value);
        }

        return hash.ToHashCode();
    }
}

File: platforms/android/mlkemnative/src/main/java/io/github/marlonjd/mlkemnative/MLKEMNative768.kt
package io.github.marlonjd.mlkemnative

import java.security.GeneralSecurityException
import java.security.ProviderException
import java.security.SecureRandom

sealed class MLKEMException(
    message: String,
    cause: Throwable? = null,
) : GeneralSecurityException(message, cause) {
    class InvalidPublicKey(message: String = "Invalid ML-KEM-768 public key") :
        MLKEMException(message)

    class InvalidPrivateKeyRepresentation(
        message: String = "Invalid ML-KEM-768 private key representation",
    ) : MLKEMException(message)

    class InvalidCiphertext(message: String = "Invalid ML-KEM-768 ciphertext") :
        MLKEMException(message)

    class InvalidIncrementalHeader(
        message: String = "Invalid ML-KEM-768 incremental header",
    ) : MLKEMException(message)

    class InvalidEncapsulationKeyVector(
        message: String = "Invalid ML-KEM-768 encapsulation key vector",
    ) : MLKEMException(message)

    class InvalidIncrementalEncapsulationSecret(
        message: String = "Invalid ML-KEM-768 incremental encapsulation secret",
    ) : MLKEMException(message)

    class RandomGenerationFailed(cause: Throwable) :
        MLKEMException("Secure random generation failed", cause)

    class OperationFailed(message: String = "ML-KEM-768 operation failed") :
        MLKEMException(message)
}

object MLKEMNative768 {
    const val PUBLIC_KEY_BYTES = 1184
    const val CIPHERTEXT_BYTES = 1088
    const val SHARED_SECRET_BYTES = 32
    const val SECRET_KEY_BYTES = 2400
    const val KEYPAIR_SEED_BYTES = 64
    const val ENCAPSULATION_SEED_BYTES = 32
    const val INCREMENTAL_HEADER_BYTES = 64
    const val ENCAPSULATION_KEY_VECTOR_BYTES = 1152
    const val CIPHERTEXT_PART1_BYTES = 960
    const val CIPHERTEXT_PART2_BYTES = 128
    const val INCREMENTAL_ENCAPSULATION_SECRET_BYTES = 64

    const val publicKeyBytes = PUBLIC_KEY_BYTES
    const val ciphertextBytes = CIPHERTEXT_BYTES
    const val sharedSecretBytes = SHARED_SECRET_BYTES
    const val incrementalHeaderBytes = INCREMENTAL_HEADER_BYTES
    const val encapsulationKeyVectorBytes = ENCAPSULATION_KEY_VECTOR_BYTES
    const val ciphertextPart1Bytes = CIPHERTEXT_PART1_BYTES
    const val ciphertextPart2Bytes = CIPHERTEXT_PART2_BYTES
    const val incrementalEncapsulationSecretBytes = INCREMENTAL_ENCAPSULATION_SECRET_BYTES

    private const val MAGIC = "KMLK1"
    private const val EK_SEED_OFFSET = ENCAPSULATION_KEY_VECTOR_BYTES
    private const val HEADER_HASH_OFFSET = ENCAPSULATION_SEED_BYTES
    private val magicBytes = MAGIC.encodeToByteArray()
    private val secureRandom = SecureRandom()

    const val PRIVATE_KEY_REPRESENTATION_BYTES =
        5 + KEYPAIR_SEED_BYTES + PUBLIC_KEY_BYTES

    class PrivateKey private constructor(
        private val secretKey: ByteArray,
        private val representationBytes: ByteArray,
        val publicKey: PublicKey,
    ) {
        /**
         * Exportable private-key representation: KMLK1 || seed64 || publicKey1184.
         * The seed portion can recreate the private key. Treat returned arrays as secret
         * material; do not log them; store them only in caller-approved protected storage;
         * clear caller-owned mutable copies when no longer needed.
         */
        val representation: ByteArray
            get() = representationBytes.copyOf()

        fun decapsulate(ciphertext: ByteArray): ByteArray {
            if (ciphertext.size != CIPHERTEXT_BYTES) {
                throw MLKEMException.InvalidCiphertext()
            }

            return PureKotlinMLKEM768.decapsulate(ciphertext.copyOf(), secretKey)
        }

        fun decapsulateParts(
            ciphertextPart1: ByteArray,
            ciphertextPart2: ByteArray,
        ): ByteArray = MLKEMNative768.decapsulateParts(this, ciphertextPart1, ciphertextPart2)

        companion object {
            fun generate(): PrivateKey {
                val seed = randomBytes(KEYPAIR_SEED_BYTES)
                try {
                    return fromSeed(seed)
                } finally {
                    seed.fill(0)
                }
            }

            /**
             * Exportable private-key representation: KMLK1 || seed64 || publicKey1184.
             * The seed portion can recreate the private key. Treat returned arrays as secret
             * material; do not log them; store them only in caller-approved protected storage;
             * clear caller-owned mutable copies when no longer needed.
             */
            fun fromRepresentation(representation: ByteArray): PrivateKey {
                if (representation.size != PRIVATE_KEY_REPRESENTATION_BYTES) {
                    throw MLKEMException.InvalidPrivateKeyRepresentation()
                }

                if (!representation.startsWithMagic()) {
                    throw MLKEMException.InvalidPrivateKeyRepresentation()
                }

                val seedStart = magicBytes.size
                val publicKeyStart = seedStart + KEYPAIR_SEED_BYTES
                val seed = representation.copyOfRange(seedStart, publicKeyStart)
                val expectedPublicKey = representation.copyOfRange(
                    publicKeyStart,
                    PRIVATE_KEY_REPRESENTATION_BYTES,
                )
                try {
                    val generated = deriveKeypair(seed)

                    if (!generated.publicKey.contentEquals(expectedPublicKey)) {
                        throw MLKEMException.InvalidPrivateKeyRepresentation()
                    }

                    return buildPrivateKey(seed, generated.publicKey, generated.secretKey)
                } finally {
                    seed.fill(0)
                }
            }

            internal fun fromSeedForTesting(seed: ByteArray): PrivateKey {
                if (seed.size != KEYPAIR_SEED_BYTES) {
                    throw MLKEMException.InvalidPrivateKeyRepresentation()
                }
                return fromSeed(seed.copyOf())
            }

            private fun fromSeed(seed: ByteArray): PrivateKey {
                val generated = deriveKeypair(seed)
                return buildPrivateKey(seed, generated.publicKey, generated.secretKey)
            }

            private fun buildPrivateKey(
                seed: ByteArray,
                publicKey: ByteArray,
                secretKey: ByteArray,
            ): PrivateKey {
                if (!PureKotlinMLKEM768.checkSecretKey(secretKey)) {
                    throw MLKEMException.OperationFailed("ML-KEM-768 secret key validation failed")
                }

                val representation = ByteArray(PRIVATE_KEY_REPRESENTATION_BYTES)
                magicBytes.copyInto(representation, 0)
                seed.copyInto(representation, magicBytes.size)
                publicKey.copyInto(representation, magicBytes.size + KEYPAIR_SEED_BYTES)

                return PrivateKey(
                    secretKey = secretKey.copyOf(),
                    representationBytes = representation,
                    publicKey = PublicKey(publicKey),
                )
            }
        }
    }

    class PublicKey(rawRepresentation: ByteArray) {
        private val rawRepresentationBytes: ByteArray = rawRepresentation.copyOf()

        val rawRepresentation: ByteArray
            get() = rawRepresentationBytes.copyOf()

        init {
            if (rawRepresentationBytes.size != PUBLIC_KEY_BYTES ||
                !PureKotlinMLKEM768.checkPublicKey(rawRepresentationBytes)
            ) {
                throw MLKEMException.InvalidPublicKey()
            }
        }

        fun encapsulate(): Encapsulation {
            val coins = randomBytes(ENCAPSULATION_SEED_BYTES)
            try {
                return encapsulateDerand(coins)
            } finally {
                coins.fill(0)
            }
        }

        internal fun encapsulateDerand(coins: ByteArray): Encapsulation {
            if (coins.size != ENCAPSULATION_SEED_BYTES) {
                throw MLKEMException.OperationFailed("Invalid ML-KEM-768 encapsulation seed")
            }

            val localCoins = coins.copyOf()
            try {
                val output = PureKotlinMLKEM768.encapsulateDerand(rawRepresentationBytes, localCoins)
                return Encapsulation(output.ciphertext, output.sharedSecret)
            } finally {
                localCoins.fill(0)
            }
        }
    }

    data class Encapsulation internal constructor(
        private val ciphertextBytes: ByteArray,
        private val sharedSecretBytes: ByteArray,
    ) {
        val ciphertext: ByteArray
            get() = ciphertextBytes.copyOf()

        val sharedSecret: ByteArray
            get() = sharedSecretBytes.copyOf()

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Encapsulation) return false
            return ciphertextBytes.contentEquals(other.ciphertextBytes) &&
                sharedSecretBytes.contentEquals(other.sharedSecretBytes)
        }

        override fun hashCode(): Int {
            var result = ciphertextBytes.contentHashCode()
            result = 31 * result + sharedSecretBytes.contentHashCode()
            return result
        }
    }

    class IncrementalPublicKey internal constructor(
        private val headerBytes: ByteArray,
        private val encapsulationKeyVectorBytes: ByteArray,
    ) {
        val header: ByteArray
            get() = headerBytes.copyOf()

        val encapsulationKeyVector: ByteArray
            get() = encapsulationKeyVectorBytes.copyOf()

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is IncrementalPublicKey) return false
            return headerBytes.contentEquals(other.headerBytes) &&
                encapsulationKeyVectorBytes.contentEquals(other.encapsulationKeyVectorBytes)
        }

        override fun hashCode(): Int {
            var result = headerBytes.contentHashCode()
            result = 31 * result + encapsulationKeyVectorBytes.contentHashCode()
            return result
        }
    }

    class IncrementalEncapsulationPart1 internal constructor(
        private val encapsulationSecretBytes: ByteArray,
        private val ciphertextPart1Bytes: ByteArray,
        private val sharedSecretBytes: ByteArray,
    ) {
        val encapsulationSecret: ByteArray
            get() = encapsulationSecretBytes.copyOf()

        val encapsSecret: ByteArray
            get() = encapsulationSecret

        val ciphertextPart1: ByteArray
            get() = ciphertextPart1Bytes.copyOf()

        val sharedSecret: ByteArray
            get() = sharedSecretBytes.copyOf()

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is IncrementalEncapsulationPart1) return false
            return encapsulationSecretBytes.contentEquals(other.encapsulationSecretBytes) &&
                ciphertextPart1Bytes.contentEquals(other.ciphertextPart1Bytes) &&
                sharedSecretBytes.contentEquals(other.sharedSecretBytes)
        }

        override fun hashCode(): Int {
            var result = encapsulationSecretBytes.contentHashCode()
            result = 31 * result + ciphertextPart1Bytes.contentHashCode()
            result = 31 * result + sharedSecretBytes.contentHashCode()
            return result
        }
    }

    fun publicKeyToIncremental(publicKey: PublicKey): IncrementalPublicKey =
        publicKeyToIncremental(publicKey.rawRepresentation)

    fun publicKeyToIncremental(publicKey: ByteArray): IncrementalPublicKey {
        val rawPublicKey = PublicKey(publicKey).rawRepresentation
        val header = ByteArray(INCREMENTAL_HEADER_BYTES)
        rawPublicKey.copyInto(
            destination = header,
            destinationOffset = 0,
            startIndex = EK_SEED_OFFSET,
            endIndex = PUBLIC_KEY_BYTES,
        )
        sha3256(rawPublicKey).copyInto(header, HEADER_HASH_OFFSET)

        return IncrementalPublicKey(
            headerBytes = header,
            encapsulationKeyVectorBytes = rawPublicKey.copyOfRange(0, EK_SEED_OFFSET),
        )
    }

    fun publicKeyFromIncremental(
        header: ByteArray,
        encapsulationKeyVector: ByteArray,
    ): ByteArray {
        if (header.size != INCREMENTAL_HEADER_BYTES) {
            throw MLKEMException.InvalidIncrementalHeader()
        }
        if (encapsulationKeyVector.size != ENCAPSULATION_KEY_VECTOR_BYTES) {
            throw MLKEMException.InvalidEncapsulationKeyVector()
        }

        val publicKey = ByteArray(PUBLIC_KEY_BYTES)
        encapsulationKeyVector.copyInto(publicKey)
        header.copyInto(
            destination = publicKey,
            destinationOffset = EK_SEED_OFFSET,
            startIndex = 0,
            endIndex = ENCAPSULATION_SEED_BYTES,
        )

        val expectedHash = sha3256(publicKey)
        val headerHash = header.copyOfRange(HEADER_HASH_OFFSET, INCREMENTAL_HEADER_BYTES)
        if (!expectedHash.contentEquals(headerHash)) {
            throw MLKEMException.InvalidIncrementalHeader(
                "Invalid ML-KEM-768 incremental header hash",
            )
        }

        PublicKey(publicKey)
        return publicKey
    }

    fun encapsulatePart1(header: ByteArray): IncrementalEncapsulationPart1 {
        if (header.size != INCREMENTAL_HEADER_BYTES) {
            throw MLKEMException.InvalidIncrementalHeader()
        }

        val coins = randomBytes(ENCAPSULATION_SEED_BYTES)
        try {
            return encapsulatePart1Derand(header, coins)
        } finally {
            coins.fill(0)
        }
    }

    @JvmSynthetic
    internal fun encapsulatePart1DerandForTesting(
        header: ByteArray,
        randomness: ByteArray,
    ): IncrementalEncapsulationPart1 {
        if (header.size != INCREMENTAL_HEADER_BYTES) {
            throw MLKEMException.InvalidIncrementalHeader()
        }
        if (randomness.size != ENCAPSULATION_SEED_BYTES) {
            throw MLKEMException.OperationFailed("Invalid ML-KEM-768 encapsulation randomness")
        }

        return encapsulatePart1Derand(header, randomness)
    }

    private fun encapsulatePart1Derand(
        header: ByteArray,
        coins: ByteArray,
    ): IncrementalEncapsulationPart1 {
        val localCoins = coins.copyOf()
        try {
            val output = PureKotlinMLKEM768.encapsulatePart1Derand(header.copyOf(), localCoins)
            return IncrementalEncapsulationPart1(
                encapsulationSecretBytes = output.encapsulationSecret,
                ciphertextPart1Bytes = output.ciphertextPart1,
                sharedSecretBytes = output.sharedSecret,
            )
        } finally {
            localCoins.fill(0)
        }
    }

    fun encapsulatePart2(
        encapsSecret: ByteArray,
        header: ByteArray,
        encapsulationKeyVector: ByteArray,
    ): ByteArray {
        if (encapsSecret.size != INCREMENTAL_ENCAPSULATION_SECRET_BYTES) {
            throw MLKEMException.InvalidIncrementalEncapsulationSecret()
        }

        val publicKey = publicKeyFromIncremental(header, encapsulationKeyVector)
        return PureKotlinMLKEM768.encapsulatePart2(encapsSecret.copyOf(), publicKey)
    }

    fun decapsulateParts(
        privateKey: PrivateKey,
        ciphertextPart1: ByteArray,
        ciphertextPart2: ByteArray,
    ): ByteArray {
        if (ciphertextPart1.size != CIPHERTEXT_PART1_BYTES) {
            throw MLKEMException.InvalidCiphertext("Invalid ML-KEM-768 ciphertext part1")
        }
        if (ciphertextPart2.size != CIPHERTEXT_PART2_BYTES) {
            throw MLKEMException.InvalidCiphertext("Invalid ML-KEM-768 ciphertext part2")
        }

        val ciphertext = ByteArray(CIPHERTEXT_BYTES)
        ciphertextPart1.copyInto(ciphertext)
        ciphertextPart2.copyInto(ciphertext, CIPHERTEXT_PART1_BYTES)
        return privateKey.decapsulate(ciphertext)
    }

    internal fun keypairDerandForTesting(seed: ByteArray): Pair<ByteArray, ByteArray> =
        deriveKeypair(seed.copyOf()).let { it.publicKey to it.secretKey }

    internal fun encapsulateDerandForTesting(
        publicKey: ByteArray,
        coins: ByteArray,
    ): Encapsulation = PublicKey(publicKey).encapsulateDerand(coins)

    private fun deriveKeypair(seed: ByteArray): NativeKeypair {
        if (seed.size != KEYPAIR_SEED_BYTES) {
            throw MLKEMException.OperationFailed("Invalid ML-KEM-768 keypair seed")
        }

        val output = PureKotlinMLKEM768.keypairDerand(seed)
        return NativeKeypair(output.publicKey, output.secretKey)
    }

    private fun randomBytes(size: Int): ByteArray {
        val output = ByteArray(size)
        try {
            secureRandom.nextBytes(output)
        } catch (exception: ProviderException) {
            throw MLKEMException.RandomGenerationFailed(exception)
        } catch (exception: RuntimeException) {
            throw MLKEMException.RandomGenerationFailed(exception)
        }
        return output
    }

    private fun ByteArray.startsWithMagic(): Boolean {
        if (size < magicBytes.size) return false
        for (index in magicBytes.indices) {
            if (this[index] != magicBytes[index]) return false
        }
        return true
    }

    private fun sha3256(input: ByteArray): ByteArray =
        PureKotlinMLKEM768.sha3256(input)

    private data class NativeKeypair(
        val publicKey: ByteArray,
        val secretKey: ByteArray,
    )

}
